src/components/DateRangeInput/index.tsx

```diff
@@ -0,0 +1,194 @@
+    { key: 'thisMonth', label: 'This month' },
+    { key: 'lastThirtyDays', label: 'Last 30 days' },
+    { key: 'lastThreeMonths', label: 'Last 3 months' },
+    { key: 'lastSixMonths', label: 'Last 6 months' },
+    { key: 'lastYear', label: 'Last year' },
+];
+
+const keySelector = (d: DateRangeOption) => d.key;
+const labelSelector = (d: DateRangeOption) => d.label;
+
+// TODO: move to hooks
+export function useInputValue<T>(initialValue: T | undefined): [
+    T | undefined,
+    (
+        v: T | undefined,
+        n?: string | undefined,
+        e?: React.FormEvent<HTMLInputElement> | React.FormEvent<HTMLTextAreaElement>,
+    ) => void
+] {
+    const [value, setValue] = React.useState<T | undefined>(initialValue);
+
+    const setInputValue = React.useCallback(
+        (newValue: T | undefined) => {
+            setValue(newValue);
+        },
+        [setValue],
+    );
+
+    return [value, setInputValue];
+}
+
+interface DateRangeModalProps {
+    onDateRangeSelect: (startDate: string, endDate: string) => void;
+    onCloseButtonClick: ModalProps['onClose'];
+}
+
+// TODO: move to separate component
+function DateRangeModal(props: DateRangeModalProps) {
+    const {
+        onDateRangeSelect,
+        onCloseButtonClick,
+    } = props;
+
+    const [startDate, setStartDate] = useInputValue('');
+    const [endDate, setEndDate] = useInputValue('');
+
+    const handleSelectButtonClick = React.useCallback(() => {
+        if (startDate && endDate) {
+            onDateRangeSelect(startDate, endDate);
+        }
+    }, [startDate, endDate, onDateRangeSelect]);
+
+    return (
+        <Modal
+            heading="Select date range"
+            className={styles.dateRangeModal}
+            onClose={onCloseButtonClick}
+            footer={(
+                <Button
+                    disabled={!startDate || !endDate}
+                    onClick={handleSelectButtonClick}
+                >
+                    Select
+                </Button>
+            )}
+        >
+            <DateInput
+                className={styles.startDateInput}
+                label="Start date"
+                value={startDate}
+                onChange={setStartDate}
+            />
+            <DateInput
+                className={styles.endDateInput}
+                label="End date"
+                value={endDate}
+                onChange={setEndDate}
+            />
+        </Modal>
+    );
+}
+
+export interface Props<N> {
+    name?: N;
+    value?: DateRangeValue;
+    onChange: (newValue: DateRangeValue | undefined, name?: N) => void;
+}
+
+// TODO: Expose more props (eg: elementRef, inputElementRef, etc)
+function DateRangeInput<N extends string>(props: Props<N>) {
+    const {
+        name,
+        value,
+        onChange,
+    } = props;
+
+    const [showCustomDateRangeModal, setShowCustomDateRangeModal] = React.useState(false);
+
+    const handleCloseButtonClick = React.useCallback(() => {
```

Comment:
Since we're using useCallback multiple times, isn't it better to import it from 'react' rather than calling React.useCallback everytime?